### DISCUSSION TITLE
conditional inclusion of a configuration file, conditional statements in config file

### DISCUSSION THREAD

**Discussion by benibilme** [i3/i3#4940] [Q&A]
TL;DR
--------
Is there any way to use conditional statements in i3 configuration file. If not how can I include certain files based on hostname at runtime via i3 capabilities.
-------

I have looked old posts in redshift and several other places and I can not seem to find a satisfying answer so I am asking this question again.

I am a newbie in i3. I am converting from dwm. I have finally reached a good enough configuration that I like and I want to use this configuration in my several other machines. 

Some machines have multiple monitors, pcs, labtops so I need change the configuration based on hostname of the computer. I have splitted the configuration file to several files based on the subject and I have also host specific configurations that are not common for all platforms. for example, I have different number of workspace configuration for each machine so that that file is machine specific.

My base configuration file is as follows which is under ~/.i3/config that is managed by gnu stow

My ~/.i3 directory is as follows

````
config
config-bar
config-binds
config-exec
config-modes
config-movements
config-scratchpad
config-workspaces
config-workspaces.pc1
config-workspaces.pc2
i3status.conf
````

config-spaces is a soft link to the corresponding machine specific file 
`config-workspaces -> config-workspaces.hostname` which may be pc1,pc2 etc. 

I set the soft link in the end of a shell script that applies stow configuration for all my dotfiles.

`ln -sr ~/.i3/config-workspaces.$(hostnamectl hostname) ~/.i3/config-workspaces`

````
font pango:monospace 12

set $mod Mod4
set $term alacritty
set $up m
set $down k
set $left t
set $right l

floating_modifier $mod

include ~/.i3/config-binds
include ~/.i3/config-movements
include ~/.i3/config-workspaces
include ~/.i3/config-modes
include ~/.i3/config-bar
include ~/.i3/config-scratchpad
include ~/.i3/config-exec
`````

I really do not like my current solution, I have bunch of other tools and only I am post processing i3 environment which I do not like. 

What I want a simple conditional inclusion in config file. I have tried to use i3-msg but, I may be wrong, but it seems i3-mgs does not honor include command such as

````
i3-msg 'include ~/.i3/config-workspace.pc1'  
ERROR: Your command: include ~/.i3/config-workspace.pc1
ERROR:               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR: Expected one of these tokens: <end>, '[', 'move', 'exec', 'exit', 'restart', 'reload', 'shmlog', 'debuglog', 'border', 'layout', 'append_layout', 'workspace', 'focus', 'kill', 'open', 'fullscreen', 'sticky', 'split', 'floating', 'mark', 'unmark', 'resize', 'rename', 'nop', 'scratchpad', 'swap', 'title_format', 'title_window_icon', 'mode', 'bar', 'gaps'
[{"success":false,"parse_error":true,"error":"Expected one of these tokens: <end>, '[', 'move', 'exec', 'exit', 'restart', 'reload', 'shmlog', 'debuglog', 'border', 'layout', 'append_layout', 'workspace', 'focus', 'kill', 'open', 'fullscreen', 'sticky', 'split', 'floating', 'mark', 'unmark', 'resize', 'rename', 'nop', 'scratchpad', 'swap', 'title_format', 'title_window_icon', 'mode', 'bar', 'gaps'","input":"include ~/.i3/config-workspace.pc1","errorposition":"^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^"}]
````

If this command did work, I would be writing a simple script which selects the correct platform dependent file and include it during i3 startup. 

Is there any way to use conditional statements in i3 configuration file. If not how can I include certain files based on hostname at runtime via i3 capabilities. 

Thank in advance.

**Comment by Airblader** [comment]
You can define a variable using `set_from_resource` which you the use for your `include` statement. X resources can be defined depending on the hostname in xrdb.

**Reply by Rom1deTroyes** [reply]
10 years ago, #759 goes to a wontfix tag for this question, so a lot of people use external tools for building one i3 config file per i3 config directory (I use ansible for that, with the same logic as yours : I build the config file by including per host conf in a base template).

Still have a great interest of making things simpler with an `include config-$(hostname)` statement from withing i3 config :1st_place_medal:

**Reply by benibilme** [reply]
> 10 years ago, #759 goes to a wontfix tag for this question, so a lot of people use external tools for building one i3 config file per i3 config directory (I use ansible for that, with the same logic as yours : I build the config file by including per host conf in a base template).
> 
> Still have a great interest of making things simpler with an `include config-$(hostname)` statement from withing i3 config 1st_place_medal

Thank you. I am glad that I am not alone. All other alternatives window managers are based on language for most of them this is easy and superior to i3.  I very much like i3 but really I do not want to use an external tool such as ansible to handle my environment. I believe include directive can be added to the i3-msg. 

I am currently using stow without bells and whistles but for .i3 I have to maintain a script as well.

**Reply by benibilme** [reply]
> You can define a variable using `set_from_resource` which you the use for your `include` statement. X resources can be defined depending on the hostname in xrdb.

Can you think of anything other than .Xresources. I am not using that file. So far, I have not found a need for it. It is another file to maintain for me.

**Reply by Airblader** [reply]
i3 has an include directive now. You can include some file that you have on all your machines (but not in stow) which simply defines a variable for the hostname. You can use that variable in your i3 config.
